### PR TITLE
Introduce soft assertions for WebTestClient

### PR DESCRIPTION
**Pull request by wyhasany**
## Overview

It happens very often that `WebTestClient` is used in heavyweight integration tests. It's no use to waste time to check if another condition has been fixed or not. Soft assertions could help a lot to check all conditions at once even if one of them fails.

## Proposal

New API would look like as follows:

```java
client.get().uri("/test")
	.exchange()
	.expectAllSoftly(
		exchange -> exchange.expectStatus().isOk(),
		exchange -> exchange.expectBody(String.class)
			.isEqualTo("It works!")
	);
```

## Related Issues

- #26917

### CODE REVIEW

File: spring-test/src/test/java/org/springframework/test/web/reactive/server/samples/SoftAssertionTests.java
Comment (cuspymd) [inline spring-test/src/test/java/org/springframework/test/web/reactive/server/samples/SoftAssertionTests.java:63]: If there are many asserts, and some of them fail, it can be difficult to determine which assert failed.
It would be nice if the indices were also included in the error message.

Comment (wyhasany) [reply spring-test/src/test/java/org/springframework/test/web/reactive/server/samples/SoftAssertionTests.java]: Resolved, pls check the changes in the code

**Review by kkocel — APPROVED** [review]

**Comment by sbrannen** [comment]
This PR is blocked until #26917 has been merged into `main`.

**Comment by sbrannen** [comment]
This has been merged into `main` in 25dca404138e85e8864e4bfa4b37c88f236a559e and revised in dd9b99e13d6b30462fb168799e6bbca1456123da.

Thanks for the 2nd _soft assertions_ PR! 👍

**Comment by rstoyanchev** [comment]
Sorry for the late review. Overall looks good, but I've spotted on follow-up issue to be fixed.

Currently individual expectations like `responseSpec.expectStatus()` are internally surrounded with `ExchangeResult#assertWithDiagnostics` which wraps any AssertionError in another `AssertionError` that adds the request and response details. That means when multiple assertions are executed, those details are also printed multiple times.

I'm not immediately sure how to solve this, but we'll need to find something.

**Comment by sbrannen** [comment]
> Currently individual expectations like `responseSpec.expectStatus()` are internally surrounded with `ExchangeResult#assertWithDiagnostics` which logs request and response details. That means when multiple assertions are executed, those details are also logged multiple times.

Good catch.

> I'm not immediately sure how to solve this, but we'll need to find something.

OK. Let's brainstorm.

**Pull request closed without merging**
